doc/api18970.md

```diff
@@ -0,0 +1,511 @@
+# Draft material to review in support of [#18970][]
+
+[//]: # (cmark-gfm -e table -e footnotes doc/api18970.md > /tmp/api18970.html)
+
+[#18970]: https://github.com/zephyrproject-rtos/zephyr/issues/18970
+
+Note: See the Background section below for technical concepts necessary
+to understand the definitions.
+
+## API Attributes
+
+This work is leading to a proposal for five specific attributes for API
+behavior that should be documented in function descriptions and, in some
+cases, exposed through API at runtime.
+
+* **reschedule** if executing the function reaches a reschedule point
+* **sleep** if executing the function can cause the invoking thread to
+  sleep
+* **no-wait** if a parameter to the function can prevent the invoking
+  thread from trying to sleep
+* **isr-ok** if the function can be safely called from an ISR even if it
+  may return an error in that case
+* **supervisor** if the calling thread must have supervisor privileges
+  to execute the function
+
+Details on the behavioral impact of each attribute are in the following
+sections.  Attributes for some common kernel API are below:
+
+name              | reschedule | sleep | no-wait | isr-ok | supervisor
+----------------- | ---------- | ----- | ------- | ------ | ----------
+`k_busy_wait`     | no         | -     | -       | -      | no
```

Comment:
`-` means it's not specified.  If something is not `reschedule`, it's not `sleep`.  If it's not `sleep`, it's not `no-wait`.  That's a consequence of the way the terms are defined. Anything that isn't `reschedule` is assumed to be callable from thread or interrupt context, with caveats.

At this stage I'm focused on the minimum number of attributes necessary to describe behavior in the case where a specific attribute is present.  We can't assume that if `driver_api_foo()` is *not* marked `reschedule` that it can't be pre-empted due to making another thread ready: maybe the way it has to be implemented requires that freedom.  Absence of attribute is not absence of behavior associated with an attribute.

Also see below on why it's not `isr-ok`.

Reply:
> If something is not `reschedule`, it's not `sleep`. If it's not `sleep`, it's not `no-wait`.

Gotcha. Isn't the proper value `no` then for `sleep` and `no-wait`, rather than `-`? From your comment, these *are* specified: the short answer is "no", the long answer is "no by definition, given the value of the `reschedule` column".

> Also see below on why it's not isr-ok

Same here. The answer seems like it's "no" rather than "-".

I would personally have found this clearer with "no" used, at least.

Reply:
I feel the distinction between "attribute is not supported because of the way the function is implemented" (and therefore the behavior could be changed to add the attribute) and "attribute is not supported because it's not relevant" is important.  People who don't understand this will have an incomplete understanding of the relationship between the attributes.

So I've marked these `no*` with `*` documenting that point.

I've also updated `isr-ok` and added a couple more attributes.  Please mark resolved any comments you've made that are no longer relevant.

Reply:
> So I've marked these `no*` with `*` documenting that point.
> 
> I've also updated `isr-ok` and added a couple more attributes.

:+1: 

>  Please mark resolved any comments you've made that are no longer relevant.

Done. The "resolve conversation" button is not available on some of my outdated comments. Please disregard all of those as of time of writing.